File: 2020/day10/day10.kt
import java.io.File

fun main() {
    val input = File("day10.txt").readLines().map { it.toInt() }.sorted()
    val input0 = listOf(0) + input
    val diff = input0.windowed(2).map { it[1] - it[0] }.groupingBy { it }.eachCount()
    println(diff.get(1)!! * (1 + diff.get(3)!!))

    val a = input0.reversed()
    var m = listOf(1L, 0L, 0L)
    for (i in 0..input.size) {
        val s = m[0] + 
            (if (i > 1 && a[i-2] - a[i] <= 3) m[1] else 0) +
            (if (i > 2 && a[i-3] - a[i] <= 3) m[2] else 0)
        m = listOf(s, m[0], m[1])
    }
    println(m[0])
}

File: 2020/day10/day10.py
from collections import Counter

a = sorted(map(int, open('day10.txt')))
c = Counter([a[i] - (a[i - 1] if i else 0) for i in range(len(a))] + [3])
print(c[1] * c[3])

b, m = list(reversed(a)) + [0], [1]
for i in range(len(a) + 1):
    s = m[-1]
    if i > 1 and b[i-2] - b[i] <= 3: s += m[-2]
    if i > 2 and b[i-3] - b[i] <= 3: s += m[-3]
    m.append(s)
print(m[-1])
